pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/service/impl/UserService.java

```diff
@@ -0,0 +1,511 @@
+        } else {
+            // if inputUser has super admin, it should return false
+            return !inputUserHasSuperAdmin;
+        }
+
+    }
+
+    @Transactional
+    public List<User> addUsers(List<User> users) {
+        SecurityContext securityContext = SecurityContextHolder.getContext();
+        CustomUserDetails customUserDetails = (CustomUserDetails) securityContext.getAuthentication().getPrincipal();
+        if (customUserDetails == null || customUserDetails.getUser() == null && customUserDetails.getUser().getUuid() == null) {
+            logger.error("Security context didn't have a user stored.");
+            return null;
+        }
+
+        User currentUser = customUserDetails.getUser();
+        checkAssociation(users);
+        for (User user : users) {
+            logger.debug("Adding User {}", user);
+            if (!allowUpdateSuperAdminRole(currentUser, user, null)) {
+                logger.error("updateUser() user - {} - with roles [{}] - is not allowed to grant " + AuthNaming.AuthRoleNaming.SUPER_ADMIN + " role when adding a user.", currentUser.getUuid(), currentUser.getRoleString());
+                throw new IllegalArgumentException("Not allowed to add a user with a " + AuthNaming.AuthRoleNaming.SUPER_ADMIN + " privilege associated.");
+            }
+
+            if (user.getEmail() == null) {
+                try {
+                    logger.info("Parsing metadata for email address");
+                    HashMap<String, String> metadata = new HashMap<String, String>(new ObjectMapper().readValue(user.getGeneralMetadata(), Map.class));
+                    List<String> emailKeys = metadata.keySet().stream().filter((key) -> key.toLowerCase().contains("email")).toList();
+                    if (!emailKeys.isEmpty()) {
+                        user.setEmail(metadata.get(emailKeys.getFirst()));
+                    }
+                } catch (IOException e) {
+                    logger.error("Failed to parse metadata for email address", e);
+                }
+            }
+        }
+
+        users = addUser(users);
+        return users;
+    }
+
+    /**
+     * check all referenced field if they are already in database. If
+     * they are in database, then retrieve it by id, and attach it to
+     * user object.
+     *
+     * @param users A list of users
+     */
+    private void checkAssociation(List<User> users) {
+        for (User user : users) {
+            if (user.getRoles() != null) {
+                Set<Role> roles = new HashSet<>();
+                user.getRoles().forEach(t -> {
+                    Optional<Role> role = this.roleService.getRoleById(t.getUuid());
+                    if (role.isEmpty()) {
+                        throw new RuntimeException("Role not found - uuid: " + t.getUuid().toString());
+                    }
+                    roles.add(role.get());
+                });
+
+                user.setRoles(roles);
+            }
```

Comment:
Wait, so for `u` users that each have `r` roles, this runs `u*r` queries? That seems bad. Why not do a `SELECT * FROM roles WHERE uuid IN (user.getRoles().stream().getID().toSet())`

Reply:
I will do this. 👍